src/euroeval/benchmark_modules/litellm.py

```diff
@@ -344,151 +361,357 @@ def generate(self, inputs: dict) -> GenerativeModelOutput:
+            (lambda: any(m in msg for m in stop_messages), disable_stop),
+            (
+                lambda: any(m in msg for m in logprobs_messages)
+                or (isinstance(error, VertexAIError) and "logprobs" in msg),
+                disable_logprobs,
+            ),
+            (lambda: any(m in msg for m in temperature_messages), disable_temperature),
+            (
+                lambda: any(m in msg for m in temperature_must_be_one_messages),
+                force_temperature_one,
+            ),
+            (lambda: any(m in msg for m in max_items_messages), disable_max_items),
+            (
+                lambda: any(m in msg for m in no_json_schema_messages),
+                fallback_json_schema,
+            ),
+            (
+                lambda: isinstance(
+                    error,
+                    (
+                        APIConnectionError,
+                        Timeout,
+                        ServiceUnavailableError,
+                        InternalServerError,
+                        SystemError,
+                    ),
+                ),
+                service_unavailable,
+            ),
+        ]
+
+        for predicate, handler in handlers:
+            if predicate():
+                handler()
+                return
+
+        if isinstance(error, RateLimitError):
+            raise InvalidModel(
+                f"You have encountered your rate limit for model {model_id!r}. "
+                "Skipping."
+            )
+
+        if isinstance(error, AuthenticationError):
+            raise NeedsAdditionalArgument(
+                cli_argument="--api-key",
+                script_argument="api_key=<your-api-key>",
+                run_with_cli=self.benchmark_config.run_with_cli,
+            )
+
+        raise InvalidBenchmark(
+            f"Failed to generate text. The error message was: {error}"
+        )
+
+    async def _generate_async(
+        self,
+        messages: list[dict[str, t.Any]],
+        generation_kwargs: dict[str, t.Any],
+        max_retries: int,
+        max_reruns: int,
+    ) -> tuple[list[ModelResponse], list[tuple[int, Exception]]]:
+        """Generate outputs from the model asynchronously.
+
+        Args:
+            messages:
+                The messages to pass to the model.
+            generation_kwargs:
+                The generation kwargs to pass to the model.
+            max_retries:
+                The maximum number of retries to make.
+            max_reruns:
+                The maximum number of reruns to make.
+
+        Returns:
+            A tuple containing the successful responses and the failed responses.
+        """
+        success = []
+        failures = []
+
+        to_run = list(enumerate(messages))
+
+        prev_fail_count = len(to_run)
+        rerun_count = 0
+
+        while to_run and rerun_count < max_reruns and prev_fail_count > 0:
+            if not to_run:
+                break
```

Comment:
This will never trigger, since `to_run` has to be non-empty for the loop to even run, no?